test-integration/test_integration/conftest.py

```diff
@@ -1,25 +1,76 @@
+from pathlib import Path
 import subprocess
 import pytest
 
-from .util import random_string, find_free_port, docker_run, get_local_ip, wait_for_port
+import redis
+
+from .util import random_string, find_free_port, docker_run, wait_for_port
 
 
 @pytest.fixture
-def redis_port():
-    container_name = "cog-test-redis-" + random_string(10)
+def docker_network():
+    name = "cog-test-" + random_string(10)
+    subprocess.run(["docker", "network", "create", name])
+    yield name
+    subprocess.run(["docker", "network", "rm", name])
+
+
+@pytest.fixture
+def redis_port(docker_network):
+    """Start a redis server inside the Docker network.
+    Inside the network, it is available at redis:6379
+    Outside the network, it is available at localhost:redis_port
+    """
     port = find_free_port()
     with docker_run(
         "redis",
-        name=container_name,
+        net_alias="redis",
+        network=docker_network,
         publish=[{"host": port, "container": 6379}],
         detach=True,
     ):
-        wait_for_port(get_local_ip(), port)
+        wait_for_port("localhost", port)
         yield port
 
 
+@pytest.fixture
+def redis_client(redis_port):
+    yield redis.Redis("localhost", redis_port)
+
+
 @pytest.fixture
 def docker_image():
     image = "cog-test-" + random_string(10)
     yield image
     subprocess.run(["docker", "rmi", "-f", image], check=False)
+
+
+@pytest.fixture
+def upload_server_image():
+    """
+    Build the upload server once for the test run. The image doesn't change.
+    """
+    subprocess.run(
+        ["docker", "build", "-t", "cog-test-upload-server", "."],
+        cwd=Path(__file__).parent.parent / "upload_server",
```

Comment:
TIL: `Path(__file__).parent.parent` -- I like how python uses dot notation for imports and paths.